core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -163,17 +163,22 @@ public SqlTaskExecution(
             else {
                 taskHandle = null;
             }
-
-            outputBuffer.addStateChangeListener(new CheckTaskCompletionOnBufferFinish(SqlTaskExecution.this));
         }
     }
 
     public void start()
     {
         try (SetThreadName ignored = new SetThreadName("Task-%s", getTaskId())) {
+            // Task handle was not created because the task is already done, nothing to do
+            if (taskHandle == null) {
+                return;
+            }
             // The scheduleDriversForTaskLifeCycle method calls enqueueDriverSplitRunner, which registers a callback with access to this object.
-            // The call back is accessed from another thread, so this code cannot be placed in the constructor.
+            // The call back is accessed from another thread, so this code cannot be placed in the constructor. This must also happen before outputBuffer
+            // callbacks are registered to prevent a task completion check before task lifecycle splits are created
             scheduleDriversForTaskLifeCycle();
+            // Output buffer state change listener callback must not run in the constructor to avoid leaking a reference to "this" across to another thread
+            outputBuffer.addStateChangeListener(new CheckTaskCompletionOnBufferFinish(SqlTaskExecution.this));
```

Comment:
nit: move to a separate commit

Reply:
Actually, this is all related code. The movement to here alone would cause issues with `scheduleDriversWithTaskLifecycle()` not being called when `taskHandle == null`. This is "fine" because it doesn't break the task state machine model today, but causes problems in a world with terminating task states.